weblab/entities/models.py

```diff
@@ -64,6 +69,126 @@ def nice_version(self, commit):
             version = version[:8] + '...'
         return version
 
+    def get_visibility_from_repo(self, commit):
+        """
+        Get the visibility of the given entity version from the repository
+
+        :param commit: `repository.Commit` object
+        :return visibility: string representing visibility
+        """
+        note = commit.get_note()
+        if note and note.startswith(VISIBILITY_NOTE_PREFIX):
+            return note[len(VISIBILITY_NOTE_PREFIX):]
+
+    def set_visibility_in_repo(self, commit, visibility):
+        """
+        Set the visibility of the given entity version in the repository
+
+        :param commit:`repository.Commit` object
+        :param visibility: string representing visibility
+        """
+        commit.add_note('%s%s' % (VISIBILITY_NOTE_PREFIX, visibility))
+
+    @property
+    def repocache(self):
+        from repocache.models import CachedEntity
+        return CachedEntity.objects.get_or_create(entity=self)[0]
+
+    def set_version_visibility(self, commit, visibility):
+        """
+        Set the visibility of the given entity version
+
+        Updates both the repository and the cache
+
+        :param commit: ref of the relevant commit
+        :param visibility: string representing visibility
+        """
+        commit = self.repo.get_commit(commit)
+        self.set_visibility_in_repo(commit, visibility)
+
+        try:
+            self.repocache.get_version(commit.hexsha).set_visibility(visibility)
+        except RepoCacheMiss:
+            pass
+
+    def get_version_visibility(self, sha):
+        """
+        Get the visibility of the given entity version
+
+        This is fetched from the repocache
+
+        :param sha: SHA of the relevant commit
+        :return: string representing visibility
+        """
+        return self.repocache.get_version(sha).visibility
```

Comment:
I've not put miss handling directly in here because I wanted the views to have freedom in how they handle this. (e.g. a 404 is appropriate in some cases). I found some other places where it wasn't handled, so I've set up a `default` arg to deal with this.